syntax = "proto3";

package whitenoise;

import "value.proto";
import "components.proto";

message Error {
    string message = 1;
}

message Analysis {
    PrivacyDefinition privacy_definition = 1;
    ComputationGraph computation_graph = 2;
}

// The definition of privacy determines parameters for sensitivity derivations and the set of available algorithms.
message PrivacyDefinition {
    // Privacy leakage with respect `group_size` number of rows. This is typically one.
    uint32 group_size = 4;

    enum Distance {
        // (e, 0)-differential privacy
        PURE = 0;

        // (e, d)-differential privacy
        APPROXIMATE = 1;
    }
    // Define how to measure distance between probability distributions.
    Distance distance = 5;

    enum Neighboring {
        SUBSTITUTE = 0;
        ADD_REMOVE = 1;
    }
    // Define what kind of perturbation may be applied to a dataset to create a neighboring dataset.
    Neighboring neighboring = 6;
}
message ComputationGraph {
    map<uint32, Component> value = 1;
}

message Release {
    map<uint32, ReleaseNode> values = 1;
}

enum FilterLevel {
    // release from runtime should include public data (either literals or sanitized data)
    PUBLIC = 0;
    // release from runtime should include public and prior known values
    PUBLIC_AND_PRIOR = 1;
    // release from runtime should include evaluations from all nodes
    ALL = 2;
}

// derived properties for the top-level Value type
message ValueProperties {
    oneof variant {
        IndexmapProperties indexmap = 1;
        ArrayProperties array = 2;
        JaggedProperties jagged = 3;
        FunctionProperties function = 4;
    }
}

// derived properties for the Value::Indexmap type
message IndexmapProperties {
    I64Null num_records = 1;
    bool disjoint = 2;
    IndexmapValueProperties value_properties = 3;

    enum Variant {
        DATAFRAME = 0;
        PARTITION = 1;
    }
    Variant variant = 4;
    I64Null dataset_id = 5;
}

message IndexmapValueProperties {
    oneof variant {
        IndexmapValuePropertiesStr string = 1;
        IndexmapValuePropertiesBool bool = 2;
        IndexmapValuePropertiesI64 i64 = 3;
    }
}

message IndexmapValuePropertiesStr {
    map<string, ValueProperties> data = 1;
}
message IndexmapValuePropertiesBool {
    map<bool, ValueProperties> data = 1;
}
message IndexmapValuePropertiesI64 {
    map<int64, ValueProperties> data = 1;
}


// sub-properties for Value::* types that may be aggregated
message AggregatorProperties {
    Component component = 1;
    map<string, ValueProperties> properties = 2;
    Array1dF64 lipschitz_constant = 3;
}

// derived properties for the Value::Array type
message ArrayProperties {
    I64Null num_records = 1;
    I64Null num_columns = 2;
    bool nullity = 3;
    Array1dF64 c_stability = 4;

    AggregatorProperties aggregator = 5;
    
    DataType data_type = 6;
    bool releasable = 7;

    oneof nature {
        NatureContinuous continuous = 100;
        NatureCategorical categorical = 101;
    }
    I64Null dataset_id = 8;
    bool is_not_empty = 9;
    I64Null dimensionality = 10;
}
message NatureContinuous {
    Array1dNull minimum = 1;
    Array1dNull maximum = 2;
}
message NatureCategorical {
    Jagged categories = 1;
}

// derived properties for the Value::Jagged type
message JaggedProperties {

    /// number of records per column
    Array1dI64 num_records = 1;
    /// true if the data may contain null values
    bool nullity = 2;
    /// set when data is aggregated, used to help compute sensitivity from the mechanisms
    AggregatorProperties aggregator = 3;
    /// either min/max or categories
    oneof nature {
        NatureContinuous continuous = 100;
        NatureCategorical categorical = 101;
    }
    /// type of data
    DataType data_type = 5;
    bool releasable = 6;
}

// derived properties for the Value::Function type
message FunctionProperties {
    bool releasable = 1;
}

// properties for each node on a graph
message GraphProperties {
    map<uint32, ValueProperties> properties = 1;
    repeated Error warnings = 2;
}

message Accuracies {
    repeated Accuracy values = 1;
}
message Accuracy {
    double value = 1;
    double alpha = 2;
}

message ComponentExpansion {
    map<uint32, Component> computation_graph = 1;
    map<uint32, ValueProperties> properties = 2;
    map<uint32, ReleaseNode> releases = 3;
    repeated uint32 traversal = 4;
}


message Function {
    ComputationGraph computation_graph = 1;
    Release release = 2;
    map<string, uint32> arguments = 3;
    map<string, uint32> outputs = 4;
}

// literals
message Value {
    oneof data {
        // bytes bytes = 1;

        // N-dimensional homogeneously typed array
        Array array = 2;

        // Key-Value pairs
        Indexmap indexmap = 3;

        // Data structure with mixed column lengths
        Jagged jagged = 4;

        // Evaluable function
        Function function = 5;
    }
}

message IndexmapStr {
    repeated string keys = 1;
    repeated Value values = 2;
}
message IndexmapBool {
    repeated bool keys = 1;
    repeated Value values = 2;
}
message IndexmapI64 {
    repeated int64 keys = 1;
    repeated Value values = 2;
}

message Indexmap {
    oneof variant {
        IndexmapStr string = 1;
        IndexmapBool bool = 2;
        IndexmapI64 i64 = 4;
    }
}

message ReleaseNode {
    Value value = 1;
    PrivacyUsages privacy_usages = 2;
    bool public = 3;
}
